timemachine/md/minimizer.py

```diff
@@ -172,15 +183,94 @@ def minimize_host_4d(mols, host_system, host_coords, ff, box, mol_coords=None) -
     final_coords = fire_minimize(x, u_impls, box, 50)
     for impl in u_impls:
         du_dx, _ = impl.execute(final_coords, box)
-        norm = np.linalg.norm(du_dx, axis=-1)
-        if not np.all(norm < MAX_FORCE_NORM):
-            raise MinimizationError(
-                f"Minimization failed to reduce large forces below threshold: |frc| = {norm} > {MAX_FORCE_NORM}"
-            )
+        check_force_norm(du_dx)
 
     return final_coords[:num_host_atoms]
 
 
+def make_gpu_impl(potentials):
+    """return bound impl of a SummedPotential constructed from potentials"""
+
+    params = [p.params for p in potentials]
+    flat_params = np.concatenate([param.reshape(-1) for param in params])
+
+    generic_potentials = [generic.from_gpu(p) for p in potentials]
+    summed_potential = generic.SummedPotential(generic_potentials, params)
+
+    return summed_potential.to_gpu().bind(flat_params).bound_impl(np.float32)
+
+
+def make_host_du_dx_fxn(mols, host_system, host_coords, ff, box, mol_coords=None):
+    """construct function to compute du_dx w.r.t. host coords, given fixed mols and box"""
+
+    # openmm host_system -> timemachine host_bps
+    host_bps, _ = openmm_deserializer.deserialize_system(host_system, cutoff=1.2)
+
+    # construct appropriate topology from (mols, ff)
+    if len(mols) == 1:
+        top = topology.BaseTopology(mols[0], ff)
+    elif len(mols) == 2:
+        top = topology.DualTopologyMinimization(mols[0], mols[1], ff)
```

Comment:
Should this be `DualTopology` rather than `DualTopologyMinimization`? The behavior looks identical (at lamb == 0, no difference if I am reading it right), but conceptual BaseTopology and DualTopologyMinimization seem different.